src/main.cpp

```diff
@@ -885,33 +884,65 @@ void MainFrame::OnTimer(wxTimerEvent &evt)
 
     if (freedvInterface.isRunning()) {
         int currentMode = freedvInterface.getCurrentMode();
+        if (currentMode != wxGetApp().m_prevMode)
+        {
+            // The receive mode changed, so the previous samples are no longer valid.
+            m_panelScatter->clearCurrentSamples();
+        }
+        wxGetApp().m_prevMode = currentMode;
+        
         if ((currentMode == FREEDV_MODE_800XA) || (currentMode == FREEDV_MODE_2400B) ) {
 
             /* FSK Mode - eye diagram ---------------------------------------------------------*/
 
             /* add samples row by row */
 
             int i;
-            for (i=0; i<g_stats.neyetr; i++) {
-                m_panelScatter->add_new_samples_eye(&g_stats.rx_eye[i][0], g_stats.neyesamp);
+            for (i=0; i<freedvInterface.getCurrentRxModemStats()->neyetr; i++) {
+                m_panelScatter->add_new_samples_eye(&freedvInterface.getCurrentRxModemStats()->rx_eye[i][0], freedvInterface.getCurrentRxModemStats()->neyesamp);
             }
         }
         else {
+            // Reset g_Nc accordingly.
+            switch(currentMode)
+            {
+                case FREEDV_MODE_1600:
+                    g_Nc = 16;
+                    m_panelScatter->setNc(g_Nc+1);  /* +1 for BPSK pilot */
+                    break;
+                case FREEDV_MODE_700C:
+                    /* m_FreeDV700Combine may have changed at run time */
+                    g_Nc = 14;
+                    if (wxGetApp().m_FreeDV700Combine) {
+                        m_panelScatter->setNc(g_Nc/2);  /* diversity combnation */
+                    }
+                    else {
+                        m_panelScatter->setNc(g_Nc);
+                    }
+                    break;
+                case FREEDV_MODE_700D:
+                case FREEDV_MODE_700E:
+                    g_Nc = 17; 
+                    m_panelScatter->setNc(g_Nc);
+                    break;
+                case FREEDV_MODE_2020:
+                    g_Nc = 31;
+                    m_panelScatter->setNc(g_Nc);
+                    break;
+            }
+            
```

Comment:
So previously we were relying on g_Nc being set by the current Tx mode?  I can see how that would mess things up, as we would be using a different g_Nc to that required for the newly chosen Rx mode.